src/JsonSchema/Metadata/Property/Factory/SchemaPropertyMetadataFactory.php

```diff
@@ -0,0 +1,268 @@
+            default => ['type' => 'string'],
+        };
+    }
+
+    /**
+     * Gets the JSON Schema document which specifies the data type corresponding to the given PHP class, and recursively adds needed new schema to the current schema if provided.
+     *
+     * Note: if the class is not part of exceptions listed above, any class is considered as a resource.
+     */
+    private function getClassType(?string $className, bool $nullable, ?bool $readableLink): array
+    {
+        if (null === $className) {
+            return ['type' => 'string'];
+        }
+
+        if (is_a($className, \DateTimeInterface::class, true)) {
+            return [
+                'type' => 'string',
+                'format' => 'date-time',
+            ];
+        }
+
+        if (is_a($className, \DateInterval::class, true)) {
+            return [
+                'type' => 'string',
+                'format' => 'duration',
+            ];
+        }
+
+        if (is_a($className, UuidInterface::class, true) || is_a($className, Uuid::class, true)) {
+            return [
+                'type' => 'string',
+                'format' => 'uuid',
+            ];
+        }
+
+        if (is_a($className, Ulid::class, true)) {
+            return [
+                'type' => 'string',
+                'format' => 'ulid',
+            ];
+        }
+
+        if (is_a($className, \SplFileInfo::class, true)) {
+            return [
+                'type' => 'string',
+                'format' => 'binary',
+            ];
+        }
+
+        if (!$this->isResourceClass($className) && is_a($className, \BackedEnum::class, true)) {
+            $enumCases = array_map(static fn (\BackedEnum $enum): string|int => $enum->value, $className::cases());
+
+            $type = \is_string($enumCases[0] ?? '') ? 'string' : 'int';
+
+            if ($nullable) {
+                $enumCases[] = null;
+            }
+
+            return [
+                'type' => $type,
+                'enum' => $enumCases,
+            ];
+        }
+
+        if (true !== $readableLink && $this->isResourceClass($className)) {
+            return [
+                'type' => 'string',
+                'format' => 'iri-reference',
+            ];
+        }
+
+        return ['type' => 'string'];
+    }
+
+    /**
+     * @param array<string, mixed> $jsonSchema
+     *
+     * @return array<string, mixed>
+     */
+    private function addNullabilityToTypeDefinition(array $jsonSchema, Type $type): array
+    {
+        if (!$type->isNullable()) {
+            return $jsonSchema;
+        }
+
+        if (\array_key_exists('$ref', $jsonSchema)) {
+            return ['anyOf' => [$jsonSchema, 'type' => 'null']];
+        }
+
+        return [...$jsonSchema, ...[
+            'type' => \is_array($jsonSchema['type'])
+                ? array_merge($jsonSchema['type'], ['null'])
+                : [$jsonSchema['type'], 'null'],
+        ]];
+    }
+}
```

Comment:
I really like this, this moves away some of the complexity of our TypeFactory and SchemaFactory